cardano-db-sync/src/Cardano/DbSync/Rollback.hs

```diff
@@ -6,109 +6,135 @@
 
-    lift $ rollbackCache cache blockId
+    deletedBlockCount <- DB.deleteBlocksBlockId trce txOutVariantType blockId epochNo (DB.pcmConsumedTxOut $ getPruneConsume syncEnv)
+    when (deletedBlockCount > 0) $ do
+      -- We use custom constraints to improve input speeds when syncing.
+      -- If they don't already exists we add them here as once a rollback has happened
+      -- we always need the constraints.
+      addConstraintsIfNotExist syncEnv trce
 
+    rollbackCache cache blockId
     liftIO . logInfo trce $ "Blocks deleted"
   where
     trce = getTrace syncEnv
     cache = envCache syncEnv
-    txOutTableType = getTxOutVariantType syncEnv
+    txOutVariantType = getTxOutVariantType syncEnv
 
 prepareRollback :: SyncEnv -> CardanoPoint -> Tip CardanoBlock -> IO (Either SyncNodeError Bool)
-prepareRollback syncEnv point serverTip =
-  DB.runDbIohkNoLogging (envBackend syncEnv) $ runExceptT action
+prepareRollback syncEnv point serverTip = do
+  DB.runDbIohkNoLogging (envDbEnv syncEnv) $ runExceptT action
   where
     trce = getTrace syncEnv
 
-    action :: MonadIO m => ExceptT SyncNodeError (ReaderT SqlBackend m) Bool
+    action :: MonadIO m => ExceptT SyncNodeError (DB.DbAction m) Bool
     action = do
       case getPoint point of
         Origin -> do
           nBlocks <- lift DB.queryCountSlotNo
           if nBlocks == 0
             then do
               liftIO . logInfo trce $ "Starting from Genesis"
+              pure True
             else do
-              liftIO
-                . logInfo trce
-                $ mconcat
+              liftIO . logInfo trce $
+                mconcat
                   [ "Delaying delete of "
                   , textShow nBlocks
                   , " while rolling back to genesis."
                   , " Applying blocks until a new block is found."
                   , " The node is currently at "
                   , textShow serverTip
                   ]
+              pure False
         At blk -> do
           nBlocks <- lift $ DB.queryCountSlotNosGreaterThan (unSlotNo $ blockPointSlot blk)
-          mBlockNo <-
-            liftLookupFail "Rollback.prepareRollback" $
-              DB.queryBlockHashBlockNo (SBS.fromShort . getOneEraHash $ blockPointHash blk)
-          liftIO
-            . logInfo trce
-            $ mconcat
-              [ "Delaying delete of "
-              , textShow nBlocks
-              , " blocks after "
-              , textShow mBlockNo
-              , " while rolling back to ("
-              , renderPoint point
-              , "). Applying blocks until a new block is found. The node is currently at "
-              , textShow serverTip
-              ]
-      pure False
+          mBlockNo <- lift $ DB.queryBlockHashBlockNo (SBS.fromShort . getOneEraHash $ blockPointHash blk)
+          case mBlockNo of
+            Nothing -> throwError $ SNErrRollback "Rollback.prepareRollback: queryBlockHashBlockNo: Block hash not found"
```

Comment:
In general the whole purpose of ExceptT is to have better "flow" and to avoid case patterns for errors. I think we lose this in many places.